Fix noisy logs in Ruby 2.5, ActiveModel 5.2 support

wflanagan opened the pull request:
This pull request fixes a duplicate column names issue we are seeing in using Cequel in ActiveJob/Sidekiq.  

Over 100k jobs, we see about 300 jobs where after the Sidekiq fork, the code errors with failure in the "finder methods"... where it has built a "find_by_id_and_id" method.  These jobs never correct, and continue to fail over and over again.  This is the same error we saw with our use of rspec-preloader in the 1.x.x branch of Cequel.  

This patch fixes it by turning the storage of column names from simple arrays into hashes.  Then, it uses the last value supplied to this as the value for production. This eliminates the ability for the column names to be duplicated on a reload.  

In addition, this pulls in the "Hornet Network" (thanks @pascalh1011) fix for noisy logs with Ruby 2.4.  I based from theirs and implemented my fixes. 

I've also incorporated "BMurray's Pull Request" for ActiveModel 5.2 support in (thanks @bmurray) to provide ActiveModel 5.2 support. 

And, lastly, I've updated a few cases of the def_delegator(s) to implement a method_missing approach to handle these cases, so we can bypass the noise of the deprecated Forwarable message. Note that in doing this, it bypasses the whole "private" thing all together by implementing the calls using method_missing.

wflanagan commented:
FYI.. this tests green on my local.

wflanagan commented:
FYI.. hold on this for a bit.. we found another case where this still has an error.

pezra commented:
@wflanagan,  thanks for improvements. i looked at the travis build and it seems that the tests are failing against cassandra 2.2 and 3.0. they pass against 2.1. the version of ruby doesn't seem to matter.

wflanagan commented:
Strange. I'm running Cassandra 3.0 locally.

wflanagan commented:
@pezra I looked at the Travis and the errors are related to lists. Given that I was trying to fix the other bug that the person mentioned with lists, I looked at that (i.e. touched that code putting puts in) but THOUGHT I took it all out and reset to a similar state.  

I'll go through the Travis logs and see if I can fix those errors.

wflanagan commented:
FYI.. i'll get back to this soon.. had to take a detour through other parts of the app for a bit..

wflanagan commented:
@pezra FYI.. this is green now.

pezra commented:
@wflanagan, 👍. Can you rebase?

wflanagan commented:
Ok.. That's my first time "rebasing" through a fork.  So, I'm not 100% sure I did that correctly.. but its on its way up... :-)

pezra commented on lib/cequel/schema/table.rb:
```diff
@@ -200,11 +256,10 @@ def compact_storage?
       protected
 
-      attr_reader :columns_by_name
+      # attr_reader :columns_by_name
```
Remove if unneeded.

pezra commented on spec/examples/record/callbacks_spec.rb:
```diff
@@ -35,12 +35,12 @@ def executed_callbacks
       self.post = post
     end
+    
+    private    
```
Does `#create_post` need to be public?

wflanagan replied:
I don't remember moving this code (maybe I did).. so I moved it back.

erkie commented:
@wflanagan @pezra this issue is really killing our logs. Can I jump in and fix these comments somehow?

pezra commented:
@erkie, please do

erkie commented:
Had a look at the code, turns out it's way above my head :(

wflanagan commented:
FYI, I've updated my original pull request to handle Ruby 2.5/2.4 noisy logs, ActiveModel 5.2, and my original fix to avoid duplicated finder methods.  I've also added a Dockerfile and docker-compose.yml so that you can do development without having to setup anything specific with Cassandra.

wflanagan commented:
FYI.. has the .has_rdoc fix in it as well.

wflanagan commented:
Checking on this.. don't want it to go bad.. :-)

erkie commented:
@pezra maybe a mention will help? :)

alxgsv commented:
I use this pull request in production for 1 month now. Thank you, @wflanagan!

wflanagan commented:
@pezra Thought I'd check/ping on this again.

pezra commented:
@wflanagan can you address the other comments from my review?

wflanagan commented:
Hey Peter,

I think I addressed most of them.  That review was more than a year ago..
and I made many of those fixes before.  At the time, it wasn't passing
certain versions of Ruby etc.  Now, it is.

Is there something else you want added/modified?

Wf


On Sun, Sep 30, 2018 at 2:00 PM Peter Williams <notifications@github.com>
wrote:

> @wflanagan <https://github.com/wflanagan> can you address the other
> comments from my review?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cequel/cequel/pull/378#issuecomment-425739152>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAQE5eiuUXg9jBtvyb9hDKbFhimOPBwBks5ugQbTgaJpZM4PWcDM>
> .
>

ilanusse commented:
Hello, is there anything I can do to help this roll along? I'm working on a Rails 5.2 project and would love to use Cequel for it.

wflanagan commented:
So, I think Peter wants me to make changes to the codebase. But it's been
MORE than a year since I submitted them. Some of the changes he requested,
I think, were fixed. But, some are still outstanding.

IF I can get a punch list of the requested changes, what's below the line
and HAS to be fixed, I guess I can try to push this out. But, a lot of what
is being asked or shown as outstanding really is in the "I have no idea"
category at this point, and I just have to run the tests and see.

And, given that we've been running this in production for quite some time,
as have others, I'm franking at this point a bit hesitant to break it by
making a bunch of items private, etc.

But, if no one else can step in and confirm these changes, I guess i'll try
to make it a priority.

Let me know.

On Fri, Oct 19, 2018 at 10:41 AM Iñaki Lanusse <notifications@github.com>
wrote:

> Hello, is there anything I can do to help this roll along? I'm working on
> a Rails 5.2 project and would love to use Cequel for it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cequel/cequel/pull/378#issuecomment-431386322>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAQE5Vof6eMbr2TEudMJQfMDE-zBIp2fks5umeSegaJpZM4PWcDM>
> .
>

pezra commented:
@wflanagan outstanding issues for this PR:

- [] https://github.com/cequel/cequel/pull/378#discussion_r155902760
- [] https://github.com/cequel/cequel/pull/378#discussion_r155902874
- [] https://github.com/cequel/cequel/pull/378#discussion_r155903209
- [] https://github.com/cequel/cequel/pull/378#discussion_r155914135

If we reach concensus on those i'll merge this.

erkie commented:
@ilanusse this PR has been open for so long that any new energy is probably appreciated. Please do!

ilanusse commented on lib/cequel/schema/table.rb:
```diff
@@ -52,24 +51,81 @@ def materialized_view?
+          column_description_store(column_desc)[column_desc.name] = column_desc
+          columns_by_name[column_desc.name]                       = column_desc
+        end
+      end
```
I believe Cassandra overwrites in this case, so let's go with last in.

wflanagan replied:
@pezra Does that suffice or do you want to do something else?

pezra replied:
I'd prefer to see this method fail on repeated columns. Repeats seem likely to be bugs, rather than intended behavior.

wflanagan replied:
Is there an existing test of this case?  FYI, the old one, I don't think, did this (failing).

pezra replied:
I doubt there is an existing test. There must not be since this branch has a green build.

wflanagan replied:
Adding a test case, and making this return an ArgumentError

ilanusse commented on lib/cequel/record/persistence.rb:
```diff
@@ -260,6 +262,16 @@ def hydrate(row)
+        raise ArgumentError, "Can't get deleter for new record" if new_record?
+        @deleter ||= Metal::Deleter.new(metal_scope)
+      end
+
```
<img width="1438" alt="screen shot 2018-10-19 at 15 59 59" src="https://user-images.githubusercontent.com/5209518/47238290-21d74780-d3b8-11e8-8ad3-9884b16f1fba.png">

Apparently, these two methods cause lots of warnings because of all the delegations being made unless they're public.

wflanagan replied:
That's correct and why I did it. Otherwise, it was a MUCH larger refactor.

wflanagan commented:
That sounds right.. that's why I moved them to public.


On Fri, Oct 19, 2018 at 3:02 PM Iñaki Lanusse <notifications@github.com>
wrote:

> *@ilanusse* commented on this pull request.
> ------------------------------
>
> In lib/cequel/record/persistence.rb
> <https://github.com/cequel/cequel/pull/378#discussion_r226750999>:
>
> > @@ -260,6 +262,16 @@ def hydrate(row)
>          self
>        end
>
> +      def updater
> +        raise ArgumentError, "Can't get updater for new record" if new_record?
> +        @updater ||= Metal::Updater.new(metal_scope)
> +      end
> +
> +      def deleter
> +        raise ArgumentError, "Can't get deleter for new record" if new_record?
> +        @deleter ||= Metal::Deleter.new(metal_scope)
> +      end
> +
>
> [image: screen shot 2018-10-19 at 15 59 59]
> <https://user-images.githubusercontent.com/5209518/47238290-21d74780-d3b8-11e8-8ad3-9884b16f1fba.png>
>
> Apparently, these two methods cause lots of warnings because of all the
> delegations being made unless they're public.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cequel/cequel/pull/378#discussion_r226750999>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAQE5b5Kyz3G-Brp5RdKgOPOQgRlmugUks5umiG-gaJpZM4PWcDM>
> .
>

pezra commented on lib/cequel/record/persistence.rb:
```diff
@@ -82,7 +82,7 @@ def hydrate(row)
       #
       def key_attributes
-        @attributes.slice(*self.class.key_column_names)
+        @cequel_attributes.slice(*self.class.key_column_names)
```
why the name change?

ilanusse replied:
When `@attributes` is exposed in Rails 5.2 it causes problems

pezra replied:
What problems does it cause? See <https://github.com/cequel/cequel/pull/395#discussion_r226765188> for detailed discussion.

pezra commented on lib/cequel/record/persistence.rb:
```diff
@@ -185,8 +185,10 @@ def loaded?(column = nil)
+        if new_record?
+          create(options)
+        else
+          update(options)
```
👍

pezra commented on lib/cequel/schema/table.rb:
```diff
@@ -13,20 +13,20 @@ class Table
       # @return [Array<DataColumn,CollectionColumn>] data columns and
       #   collection columns defined on the table
-      attr_reader :data_columns
+      # attr_reader :data_columns
```
let's delete these if they aren't needed anymore.

wflanagan replied:
ok.. i'll push this change.

ilanusse commented:
Created a new PR (#397) because @wflanagan mentioned having production sites running on his fork code, so I didn't want to mess around with his fork.

wflanagan commented on lib/cequel/schema/table.rb:
```diff
@@ -52,24 +51,81 @@ def materialized_view?
+          Cequel::Schema::ClusteringColumn => clustering_columns_hash
+        }.try(:[], column_desc.class)
+        value = data_columns_hash if value.nil?
+        value
```
changed in my local branch.. will push in a bit.

pezra commented on lib/cequel/record/data_set_builder.rb:
```diff
@@ -122,6 +122,14 @@ def set_paging_state
+        else
+          super
+        end
+      end
```
`#method_missing` is *much* slower than straight delegation. Why change?

pezra replied:
If the problem was the warning about delegating to a private method we could convert the target method to `protected`. Protected is generally a better choice than private anyway, IMO.

wflanagan replied:
For the schema/table.rb note, it now fails on repeated columns. I added a test to test this case as well.

wflanagan replied:
Would have been great to get this feedback about 8 months ago. :-) Alright.. i'll rework the entire approach to try to make it work.  Ugh. :-(

pezra replied:
Sorry i didn't see this until today. I changes i did request didn't get done until today, though.

If we need access to those method from another class then they should be public. Period. Use `def_delegate` to bypass method access control is a terrible idea. Let's just make them public and call it a day.

wflanagan replied:
I guess this boils down to a difference in our respective development cultures. 

I've already, based on your last comment, reimplemented them in a way that doesn't call the method_missing at all or use delegation, by exposing them as method directly on the classes.  I had already started this before I saw your comment above (stayed up late last night to implement this to get this finally done). 

If this is acceptable, then OK. If not, let me know.. as I'm not super excited about making another significant change to the code base. 

On your other comments about changes not being done until yesterday, a few thoughts for the future. 

With all the other changes made, and that **I did go through and remove comments**, but evidently missed a few, I wouldn't have expected that to hold up these changes for as long as it has.  I was under the impression that I had removed all the commented out code. If you were waiting on me to remove the 3 or 4 more lines of code, then it would have been great for you to tell me that, and point me to where you found them for me to remove. Or, even better, just remove them yourself. That "lag" of 8 months over 3 or 4 lines of commented out code seems like a waste.

And, on the other issue (add_column erroring versus taking the last one in), I didn't perceive your comment as an absolute request--more of a question about how it was implemented. We're talking about a method that had 3 or 4 lines of code. If there was a needed change, it wouldn't have been hard to fix.. and all of this could have been in production 8+ months ago. 
Anyway, it's been changed to raise an error now, and I added a test for it to ensure that it does. 

Anyway, it's different culture, and all good. 

Let me know about the delegation stuff.

pezra commented on lib/cequel/record/lazy_record_collection.rb:
```diff
@@ -77,12 +74,21 @@ def assert_fully_specified!
+        %i[table connection key_columns scoped_key_values]
+      end
+
+      def method_missing(m, *args, &block)
```
`#method_missing` is much slower than straight delegation. Why change?

wflanagan replied:
The delegation is to a private method, and this throws off MASSIVE logs. This is a part of this change. If it's reimplemented as a delegation, then it will start to throw a LOT of messages about delegating to a private method.  

This approach keeps the messages from happening, and still raises on the non-defined methods. 

I pulled this approach from another gem that had the same problem... this approach was how they solved it.

pezra replied:
Let's just make the target methods public. See https://github.com/cequel/cequel/pull/378#discussion_r226803027

wflanagan commented:
FYI.. i'm pushing a version momentarily that fixes these comments.

wflanagan commented:
@pezra @ilanusse Just pushed fixes for all requested items. I also fixed the Fixnum deprecation warning on Ruby 2.5.

wflanagan commented:
Alright. I won't get to this today.. and not sure when.

wflanagan commented:
@pezra I've tested this, and it seems that it still throws the same warning logs using protected as it does with private...

wflanagan commented:
@pezra Alright, I did it a different way that avoids using method_missing. Please take a look and see if that way works better for you.  

The change to protected DID NOT solve the problem.  The messages still print in Ruby 2.4+. So, instead, I created methods explicitly that were the ones being delegated.  

This might perform better. But, IMO, the code isn't as maintainable.. and I can think of some edge cases (related to blocks) where things might break. However, evidently, in Rails, this is how they were handling the issue for backward compatibility.  So, maybe that's a good solution for you? 

Or, would you like something done another way?

pezra commented:
Closing in favor of #401
